ocp_resources/utils.py

```diff
@@ -26,76 +31,178 @@ class TimeoutSampler:
+
+        Validate exception inputs:
+            exceptions
+            exceptions_msg
+            exceptions_dict
+        """
+        output = {}
+        if exceptions_dict and (exceptions or exceptions_msg):
+            raise InvalidArgumentsError(
+                "Must specify either exceptions_dict or exceptions/exception_msg, not both"
+            )
+
+        elif exceptions or exceptions_msg:
+            warn(
+                "TimeoutSampler() exception and exception_msg are now deprecated. "
+                "Please update to use exceptions_dict by 2021-MM-DD",
+                DeprecationWarning,
+            )
+
+        if exceptions_dict:
+            output.update(exceptions_dict)
+
+        elif exceptions is None:
+            output[Exception] = []
+
+        else:
+            if not isinstance(exceptions, tuple):
+                exceptions = (exceptions,)
+
+            for exp in exceptions:
+                if exceptions_msg:
+                    output[exp] = [exceptions_msg]
+                else:
+                    output[exp] = []
+
+        return output
+
     @property
     def _func_log(self):
         return f"Function: {self.func} Kwargs: {self.func_kwargs}"
 
     def __iter__(self):
-        last_exp = None
         timeout_watch = TimeoutWatch(timeout=self.wait_timeout)
         if self.print_log:
             LOGGER.info(
                 f"Waiting for {self.wait_timeout} seconds, retry every {self.sleep} seconds"
             )
 
+        last_exp_log = "No exceptions raised"
         while timeout_watch.remaining_time() > 0:
             try:
                 self.elapsed_time = self.wait_timeout - timeout_watch.remaining_time()
                 yield self.func(**self.func_kwargs)
                 self.elapsed_time = None
+
                 time.sleep(self.sleep)
 
-            except self.exception as exp:
-                last_exp = exp
-                self.elapsed_time = None
+            except self._exceptions as exp:
+                last_exp_log = self._get_exception_log(exp=exp)
+                if self._is_raisable_exception(exp=exp):
+                    LOGGER.error(last_exp_log)
+                    raise exp
+                else:
+                    LOGGER.warning(f"{last_exp_log}: Retrying")
 
+                self.elapsed_time = None
                 time.sleep(self.sleep)
 
             finally:
                 if self.elapsed_time and self.print_log:
                     LOGGER.info(f"Elapsed time: {self.elapsed_time}")
 
-        raise TimeoutExpiredError(self._process_execution(exp=last_exp))
+        raise TimeoutExpiredError(last_exp_log)
 
-    def _process_execution(self, exp=None):
-        exp_name = exp.__class__.__name__ if exp else "N/A"
-        last_exception_log = f"Last exception: {exp_name}: {exp}"
+    @staticmethod
+    def _is_exception_message_matched(exp, exception_messages):
+        if not exception_messages:
```

Comment:
`_is_exception_message_matched` implies that something is looked for.
It should be called only if there're `exception_messages`

Reply:
https://github.com/RedHatQE/openshift-python-wrapper/pull/24#discussion_r681664442